kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt

```diff
@@ -902,6 +929,84 @@ private fun ObjCExportMapper.canHaveSameName(first: PropertyDescriptor, second:
     return bridgePropertyType(first) == bridgePropertyType(second)
 }
 
+private class ObjCName(
+        private val kotlinName: String,
+        private val objCName: String?,
+        private val swiftName: String?,
+        val isExact: Boolean
+) {
+    // TODO: Prevent mangling when objCName or swiftName is provided
+
+    fun asString(forSwift: Boolean): String = swiftName.takeIf { forSwift } ?: objCName ?: kotlinName
+
+    fun asIdentifier(forSwift: Boolean, default: () -> String = kotlinName::toIdentifier): String =
+            swiftName.takeIf { forSwift } ?: objCName ?: default()
+}
+
+private fun DeclarationDescriptor.getObjCName(): ObjCName {
+    var objCName: String? = null
+    var swiftName: String? = null
+    var isExact = false
+    annotations.findAnnotation(KonanFqNames.objCName)?.let { annotation ->
+        objCName = annotation.argumentValue("name")?.value?.cast()
+        swiftName = annotation.argumentValue("swiftName")?.value?.cast()
+        isExact = annotation.argumentValue("exact")?.value?.cast() ?: false
+    }
+    return ObjCName(name.asString(), objCName, swiftName, isExact)
+}
+
+private fun CallableDescriptor.getObjCName(): ObjCName =
+        overriddenDescriptors.firstOrNull()?.getObjCName() ?: (this as DeclarationDescriptor).getObjCName()
```

Comment:
Can we do this without casts? (also below).
Motivation: Although these are cast to supertypes, the compiler doesn't check that. Now, imagine we are going to rewrite this code to a different representation for declarations. If we forget to change a target type in a particular cast, and we don't have tests covering this case, the compiler won't help us to detect this either, and this bug will remain unnoticed.

Let me propose alternatives:
1. Implement and use a special generic `upcast` function that doesn't do `as`, but simply returns the argument as its supertype.
2. Use different names for "conflicting" `getObjCName` functions.

Reply:
I actually like that the functions have the same name (and naming is hard), so I went with the `upcast` function.
Not sure if this is a common use case, but I guess `upcast` could even be part of stdlib.